ncm-download/src/main/perl/download.pm

```diff
@@ -1,247 +1,265 @@
-            if (!$success) {
-                $self->warn("failed to retrieve $source from any proxies, using original");
+        if ($source !~ m{^//}) {
+            if ($file->{server}) {
+                $source = "//$file->{server}$source";
+            } else {
+                # TODO: ignoring with an error?
+                #       better would be to force this being non-empty via schema
+                #       (eg no empty server attribute; or no empty server without default server specified),
+                #       and not support ignoring files with error reported
+                $self->error("$fn requested but no server, ignoring");
+                return;
             }
         }
-        if (!$success) {
-            if (!$self->download(
-                    file => $file,
-                    href => $source,
-                    timeout => $inf->{timeout},
-                    gssneg => $gss,
-                    cacert => $inf->{files}->{$f}->{cacert},
-                    capath => $inf->{files}->{$f}->{capath},
-                    cert => $inf->{files}->{$f}->{cert},
-                    key =>  $inf->{files}->{$f}->{key},
-                    min_age => $min_age)) {
-                $self->error("failed to retrieve $source, skipping");
-                next;
+
+        if ($source =~ m{^//}) {
+            # server specified, but no proto
+            if ($file->{proto}) {
+                $source = "$file->{proto}:$source";
+            } else {
+                # TODO: ignoring with error?
+                $self->error("$fn requested but no proto, ignoring");
+                return;
             }
         }
-        my @opts = ();
-        if (exists $inf->{files}->{$f}->{perm}) {
-            push(@opts, 'mode' => $inf->{files}->{$f}->{perm});
-        }
-        if (exists $inf->{files}->{$f}->{owner}) {
-            push(@opts, 'owner' => $inf->{files}->{$f}->{owner});
+    }
+
+    my %opts = (
+        href => $source,
+        gss_ccache => $file->{gss_ccache},
+        cacert => $file->{cacert},
+        capath => $file->{capath},
+        cert => $file->{cert},
+        key =>  $file->{key},
+        min_age => $file->{min_age},
+        # TODO: head_timeout used to be only per file for proxies, and per component for non-proxy
+        head_timeout => $file->{head_timeout},
+        # TODO: timeout used to be only per file for proxies, and not at all for non-proxy
+        timeout => $file->{timeout},
+        );
+
+    my $success = 0;
+    if (@$proxyhosts && $file->{proxy}) {
+        my $attempts = scalar @$proxyhosts;
+        while ($attempts--) {
+            # take the head off the list and rotate to the end...
+            # we do this in order to avoid continually trying a dead
+            # proxy - once we've found one that works, we'll keep using it.
+            my $proxy = shift @$proxyhosts;
+            $self->debug(3, "Trying proxy $proxy for source $source and filename $fn");
```

Comment:
$proxyhosts seems to be getting more deeply nested with each iteration

```
print Dumper($proxyhosts);
$VAR1 = [
          [
            [
              [
                [
                  [
                    [
                      'p3.x.x.x',
                      'p4.x.x.x',
                      'p5.x.x.x',
                      'p6.x.x.x'
                    ]
                  ]
                ]
              ]
            ]
          ]
        ];
```

but if I call Dumper just after this line:
```
    my $proxyhosts = $tree->{proxyhosts} || [];
```

it's nested correctly. I think the problem code is:

```
            if ($success) {
                @$proxyhosts = [$proxy, @$proxyhosts];
                last;
            } else {
                @$proxyhosts = [@$proxyhosts, $proxy];
            }
```
The move from an array to arrayref seems to have broken this construct. Changing to:
```
            if ($success) {
                #@$proxyhosts = [$proxy, @$proxyhosts];
                unshift  @$proxyhosts, $proxy;
                last;
            } else {
                #@$proxyhosts = [@$proxyhosts, $proxy];
                push @$proxyhosts, $proxy;
            }
```
works on RH6.